### PR TITLE
Add ConfigUnmarshaler interface to allow mutations on the parsed Config

### PR DESCRIPTION
**Pull request by bogdandrutu**
Fixes https://github.com/open-telemetry/opentelemetry-collector/issues/3023

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

### CODE REVIEW

**Review by carlosalberto — APPROVED** [review]

File: service/collector.go [hunk 1]
```diff
@@ -99,14 +100,26 @@ func New(set CollectorSettings) (*Collector, error) {
...
 
+	if col.parserProvider == nil {
+		// use default provider.
+		col.parserProvider = parserprovider.Default()
+	}
+
+	if col.configLoader == nil {
+		// use default provider.
+		col.configLoader = configloader.NewDefaultLoader()
+	}
```
Comment (tigrannajaryan) [inline service/collector.go:121]: Can we document the relationship between ConfigLoader and ParserProvider? I think it is becoming a bit confusing. A documentation explaining how config works in general would help a lot.
Can be a future PR.

Comment (bogdandrutu) [reply service/collector.go:121]: I will do in a future PR, since ConfigSource may also be confusing. I will try to explain that better.

**Comment by tigrannajaryan** [comment]
Build failed somewhere in config tests, not sure if related to the change.

File: service/settings.go [hunk 2]
```diff
@@ -60,6 +61,10 @@ type CollectorSettings struct {
 	// properties supplied via --set command line flag.
 	ParserProvider parserprovider.ParserProvider
 
+	// ConfigLoader loads the configuration's Parser into the service configuration.
+	// If it is not provided a default provider is used.
```
Comment (pjanotti) [inline service/settings.go:65]: nit:
```suggestion
	// ConfigLoader unmarshalls the configuration's Parser into the service configuration.
	// If it is not provided a default loader is used.
```

**Review by pjanotti — APPROVED** [review]
When I first read the name ConfigLoader I thought that it was related to load the configuration file. The "loader" unmarshall a config parser into a service configuration. Perhaps the name should be something around config builder or unmarshall, load sounds to me more about "loading" the configuration into the parser.

**Comment by bogdandrutu** [comment]
@pjanotti good feedback, will update the PR.

File: service/settings.go (hunk 2 above)
Comment (bogdandrutu) [reply service/settings.go]: Done.

File: service/collector.go [hunk 3]
```diff
@@ -79,6 +79,7 @@ type Collector struct {
 	factories component.Factories
 
 	parserProvider parserprovider.ParserProvider
+	configLoader   configunmarshaler.ConfigUnmarshaler
```
Comment (pjanotti) [inline service/collector.go:82]:
```suggestion
	configUnmarshaler   configunmarshaler.ConfigUnmarshaler
```

File: service/collector.go [hunk 4]
```diff
@@ -99,14 +100,26 @@ func New(set CollectorSettings) (*Collector, error) {
...
 
 	col := &Collector{
-		info:         set.BuildInfo,
-		factories:    set.Factories,
-		stateChannel: make(chan State, Closed+1),
+		info:           set.BuildInfo,
+		factories:      set.Factories,
+		stateChannel:   make(chan State, Closed+1),
+		parserProvider: set.ParserProvider,
+		configLoader:   set.ConfigLoader,
```
Comment (pjanotti) [inline service/collector.go:107]: Are you keeping `set.ConfigLoader` to avoid a breaking change on the struct?

**Review by pjanotti — APPROVED** [review]

**Review by pavolloffay — APPROVED** [review]

File: service/collector.go (hunk 3 above)
Comment (bogdandrutu) [reply service/collector.go]: Done.

File: service/collector.go (hunk 4 above)
Comment (bogdandrutu) [reply service/collector.go]: Done.

**Comment by bogdandrutu** [comment]
@tigrannajaryan @jrcamp @dmitryax can one of you approve this?

**Review by tigrannajaryan — APPROVED** [review]
Approved. Looking forward to see docs in a future PR.

**Comment by bogdandrutu** [comment]
@jcchavezs I think even though this may not seem as intuitive, this design choice gives us the best of both worlds. The reason is because it is generic enough to allow us to change the unmarshalling logic as well as being able to change the generated config.

I do understand your concern, but I would say that your proposal is very focused on your use-case, and may be too specific, so instead I believe that going with a more generic approach like this allows us to extend in the future everything that we need. There may be implementations that need to use other structure tags, or that may override some values, etc.

**Pull request merged**